UCLA Radio/Custom Elements/ChatView.swift

```diff
@@ -1,232 +1,266 @@
-//        
-//        return constraints
-//    }
-//    
-//}
+
+import Foundation
+import UIKit
+import SocketIO
+
+class ChatView: UIView, UITextViewDelegate, UITableViewDelegate, UITableViewDataSource{
+    
+    
+    var chatBox: UITextView!
+    
+    var sendButton: UIButton!
+    
+    var tableView: UITableView!
+    
+    var messagesArray: [ChatBubbleData]!
+    
+    var socketClient: SocketIOClient!
+    
+    var manager:SocketManager!
+    var socket:SocketIOClient!
+    
+    var username: String = "DEFAULT"
+    
+    var lastID: Int = 0
+    
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        
+        backgroundColor = Constants.Colors.darkBackground
+        
+        messagesArray = [ChatBubbleData]()
+        
+        
+        //declare our various views to be used in this view controller
+        chatBox = UITextView(frame: CGRect(x: 0, y: 0, width: self.frame.width*3/5, height: 20))
+        chatBox.text = "Write a message"
+        chatBox.textColor = UIColor.lightGray
+        chatBox.delegate = self
+        chatBox.backgroundColor = UIColor.clear
+        chatBox.font = UIFont(name: (chatBox.font?.fontName)!, size: 20)
+        
+        sendButton = UIButton(type: .roundedRect)
+        sendButton.setTitle("Send", for: UIControlState.normal)
+        sendButton.titleLabel?.font = UIFont(name: (sendButton.titleLabel?.font.fontName)!, size: 20)
+        sendButton.sizeToFit()
+        sendButton.addTarget(self, action: #selector(sendButtonAction), for: .touchUpInside)
+        
+        tableView = UITableView(frame: CGRect(x: 0, y: 0, width: self.frame.width, height: self.frame.height*4/5))
+        self.tableView.separatorStyle = .none
+        tableView.backgroundColor = UIColor.clear
+        tableView.register(ChatBubbleCell.self, forCellReuseIdentifier: "cell")
+        tableView.dataSource = self
+        tableView.delegate = self
+        //This line flips the y-axis of our table view, basically flipping it upside down.
+        //We do this because it allows us to have chat bubbles appear from the bottom up, not up to bottom.
+        tableView.transform = CGAffineTransform(scaleX: 1, y: -1)
+        
+        addSubview(chatBox)
+        addSubview(tableView)
+        addSubview(sendButton)
+        
+        chatBox.translatesAutoresizingMaskIntoConstraints = false;
+        sendButton.translatesAutoresizingMaskIntoConstraints = false;
+        //tableView.translatesAutoresizingMaskIntoConstraints = false;
+
+        
+        //socketIO stuff.  First we create a manager object for our connection
+        //then we add event handlers to handle events such as assign username
+        //and new messages, then we connect to the server
+        self.manager = SocketManager(socketURL: URL(string: "https://uclaradio.com")!, config: [.log(true),.compress])
```

Comment:
so this is just defining what server to refer to (or what the "manager" is, in SocketIO lingo?) right? the force unwrapping of the URL element won't cause any bugs you're aware of later?

Reply:
Yeah the manager is just an object that actually handles the socket we're using to communicate with the server.  I think the force unwrapping of the URL element is just insofar as the conversion from string to URL could potentially fail with certain strings, so an optional value is returned by this expression, however since we only use this one fixed URL I don't think this force unwrapping should ever fail.